File: src/Niche/Prolog/facade.pl
auth_service(User, Result) :-
    format(string(Result), 'auth(~w)', [User]).

inventory_service(Sku, Result) :-
    format(string(Result), 'reserve(~w)', [Sku]).

billing_service(Amount, Result) :-
    format(string(Result), 'charge(~w)', [Amount]).

checkout_facade(User, Sku, Amount, Result) :-
    auth_service(User, Auth),
    inventory_service(Sku, Inventory),
    billing_service(Amount, Billing),
    format(string(Result), '~w>~w>~w', [Auth, Inventory, Billing]).

main :-
    checkout_facade(alice, 'SKU-42', 499, Result),
    format('checkout=~w~n', [Result]).

File: src/Niche/Nim/facade.nim
type
  AuthService = object
  InventoryService = object
  BillingService = object
  CheckoutFacade = object
    auth: AuthService
    inventory: InventoryService
    billing: BillingService

proc authenticate(_: AuthService; user: string): string =
  "auth(" & user & ")"

proc reserve(_: InventoryService; sku: string): string =
  "reserve(" & sku & ")"

proc charge(_: BillingService; amount: int): string =
  "charge(" & $amount & ")"

proc checkout(facade: CheckoutFacade; user, sku: string; amount: int): string =
  "checkout=" & facade.auth.authenticate(user) & ">" &
    facade.inventory.reserve(sku) & ">" & facade.billing.charge(amount)

let facade = CheckoutFacade(
  auth: AuthService(),
  inventory: InventoryService(),
  billing: BillingService(),
)

echo facade.checkout("alice", "SKU-42", 499)
